modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileReceiver.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.network.file.exception.FileTransferException;
+import org.apache.ignite.internal.network.file.messages.FileChunkMessage;
+import org.apache.ignite.internal.network.file.messages.FileHeaderMessage;
+import org.apache.ignite.internal.network.file.messages.FileTransferErrorMessage;
+import org.apache.ignite.internal.network.file.messages.FileTransferInfoMessage;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.internal.util.RefCountedObjectPool;
+
+/**
+ * File receiver.
+ */
+class FileReceiver implements ManuallyCloseable {
+    private static final IgniteLogger LOG = Loggers.forClass(FileReceiver.class);
+
+    private final ExecutorService executorService;
+
+    private final Map<UUID, FileTransferMessagesHandler> transferIdToHandler = new ConcurrentHashMap<>();
+
+    private final Map<String, Set<UUID>> senderConsistentIdToTransferIds = new ConcurrentHashMap<>();
+
+    private final RefCountedObjectPool<String, ReentrantLock> senderConsistentIdToLock = new RefCountedObjectPool<>();
+
+    FileReceiver(String nodeName, int threadPoolSize) {
+        this.executorService = new ThreadPoolExecutor(
+                0,
+                threadPoolSize,
+                0L, TimeUnit.MILLISECONDS,
+                new LinkedBlockingQueue<Runnable>(),
+                NamedThreadFactory.create(nodeName, "file-receiver", LOG)
+        );
+    }
+
+    FileTransferMessagesHandler registerTransfer(String senderConsistentId, UUID transferId, Path handlerDir) {
+        ReentrantLock lock = senderConsistentIdToLock.acquire(senderConsistentId, ignored -> new ReentrantLock());
+        lock.lock();
+        try {
+            FileTransferMessagesHandler handler = new FileTransferMessagesHandler(handlerDir);
+            transferIdToHandler.put(transferId, handler);
+            senderConsistentIdToTransferIds.compute(senderConsistentId, (k, v) -> {
+                if (v == null) {
+                    v = new HashSet<>();
+                }
+                v.add(transferId);
+                return v;
+            });
+            handler.result().whenComplete((files, throwable) -> deregisterTransfer(senderConsistentId, transferId));
+            return handler;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void deregisterTransfer(String senderConsistentId, UUID transferId) {
+        ReentrantLock lock = senderConsistentIdToLock.acquire(senderConsistentId, ignored -> new ReentrantLock());
+        lock.lock();
+        try {
+            transferIdToHandler.remove(transferId);
+            Set<UUID> uuids = senderConsistentIdToTransferIds.get(senderConsistentId);
+            uuids.remove(transferId);
+            if (uuids.isEmpty()) {
+                senderConsistentIdToTransferIds.remove(senderConsistentId);
+            }
```

Comment:
In previous method you use `compute`, but here you use get/remove. What's the reason for discrepancy?